src/server/scripts/Kalimdor/CavernsOfTime/CullingOfStratholme/culling_of_stratholme.h

```diff
@@ -18,86 +18,132 @@
+    DATA_NOTIFY_DEATH,     // sent by wave mob AI to instance script on death
 
-    NPC_TIME_RIFT        = 28409,
-    NPC_GUARDIAN_OF_TIME = 32281
+    // old stuff below this, need to figure out if needed
+    DATA_ARTHAS
 };
 
-enum GameObjectIds
+// these are sent by instance AI to creatures; they are passed as negative values to avoid conflicts with creature script specific actions
+enum InstanceActions
 {
-    GO_SHKAF_GATE       = 188686,
-    GO_MALGANIS_GATE_1  = 187711,
-    GO_MALGANIS_GATE_2  = 187723,
-    GO_EXIT_GATE        = 191788,
-    GO_MALGANIS_CHEST_N = 190663,
-    GO_MALGANIS_CHEST_H = 193597,
-    GO_SUSPICIOUS_CRATE = 190094,
-    GO_PLAGUED_CRATE    = 190095
+    ACTION_PROGRESS_UPDATE = 1,
+    ACTION_REQUEST_NOTIFY,
+    ACTION_CORRUPTOR_LEAVE
 };
 
-enum WorldStatesCoT
+enum InstanceMisc
 {
     WORLDSTATE_SHOW_CRATES          = 3479,
     WORLDSTATE_CRATES_REVEALED      = 3480,
     WORLDSTATE_WAVE_COUNT           = 3504,
+    WORLDSTATE_WAVE_MARKER_ES       = 3581,
+    WORLDSTATE_WAVE_MARKER_FL       = 3582,
+    WORLDSTATE_WAVE_MARKER_KS       = 3583,
+    WORLDSTATE_WAVE_MARKER_MR       = 3584,
+    WORLDSTATE_WAVE_MARKER_TH       = 3585,
     WORLDSTATE_TIME_GUARDIAN        = 3931,
     WORLDSTATE_TIME_GUARDIAN_SHOW   = 3932
 };
 
-enum CrateSpells
-{
-    SPELL_CRATES_CREDIT     = 58109
-};
+// These methods are implemented in instance_culling_of_stratholme.cpp so they can access internals of the instance script
+void StratholmeAIHello(InstanceScript* instance, ObjectGuid const& me, ProgressStates myStates);
+void StratholmeAIGoodbye(InstanceScript* instance, ObjectGuid const& me, ProgressStates myStates);
 
-enum Texts
+// Note: instance script is not nullptr checked in this AI - ONLY use this with GetInstanceAI in script AI getter!
+template <class ParentAI>
+class StratholmeNPCAIWrapper : public ParentAI
 {
-    SAY_CRATES_COMPLETED    = 0,
-    // Chromie
-    SAY_INFINITE_START      = 0, // On Infinite Corruptor event start
-    SAY_INFINITE            = 1, // On Infinite Corruptor event at 5 minutes
-    SAY_INFINITE_FAIL       = 2, // On Infinite Corruptor event fail
-    // Infinite Corruptor
-    SAY_FAIL_EVENT          = 2 // On Infinite Corruptor event fail
-};
+    public:
+        StratholmeNPCAIWrapper(Creature* creature, ProgressStates stateMask) : ParentAI(creature), instance(creature->GetInstanceScript()), _statesMask(stateMask), _deathNotify(false)
+        {
+            StratholmeAIHello(instance, this->me->GetGUID(), _statesMask);
+            CheckDespawn();
+        }
+        ~StratholmeNPCAIWrapper()
+        {
+            StratholmeAIGoodbye(instance, this->me->GetGUID(), _statesMask);
+        }
 
-enum InstanceEvents
-{
-    EVENT_INFINITE_TIMER    = 1
-};
+        void CheckDespawn()
+        {
+            if (!(_statesMask & instance->GetData(DATA_INSTANCE_PROGRESS)))
+                me->DespawnOrUnsummon(0, Seconds(1));
```

Comment:
/home/travis/build/TrinityCore/TrinityCore/src/server/scripts/Kalimdor/CavernsOfTime/CullingOfStratholme/culling_of_stratholme.h:110:17: fatal error: 
      use of undeclared identifier 'me'
                me->DespawnOrUnsummon(0, Seconds(1));
                ^

Realy man, Travis dont like you xD